src/backend/gl/src/lib.rs

```diff
@@ -32,34 +41,60 @@ mod queue;
 mod state;
 mod window;
 
-#[cfg(all(not(target_arch = "wasm32"), feature = "glutin"))]
-pub use crate::window::glutin::{config_context, Headless, Surface, Swapchain};
+// Web implementation
 #[cfg(target_arch = "wasm32")]
 pub use window::web::{Surface, Swapchain};
 
-#[cfg(all(feature = "wgl", not(target_arch = "wasm32")))]
-use window::wgl::DeviceContext;
+// Glutin implementation
+#[cfg(feature = "glutin")]
+pub use crate::window::glutin::{Instance, Surface, Swapchain};
 
-#[cfg(all(feature = "wgl", not(target_arch = "wasm32")))]
-pub use window::wgl::{Surface, Swapchain, Instance};
+// Surfman implementation
+#[cfg(feature = "surfman")]
+pub use crate::window::surfman::{Instance, Surface, Swapchain};
 
-#[cfg(not(any(target_arch = "wasm32", feature = "glutin", feature = "wgl")))]
+// WGL implementation
+#[cfg(feature = "wgl")]
+use window::wgl::DeviceContext;
+#[cfg(feature = "wgl")]
+pub use window::wgl::{Instance, Surface, Swapchain};
+
+// Catch-all dummy implementation
+#[cfg(not(any(
+    target_arch = "wasm32",
+    feature = "glutin",
+    feature = "surfman",
+    feature = "wgl"
+)))]
 pub use window::dummy::{Surface, Swapchain};
 
 pub use glow::Context as GlContext;
 use glow::HasContext;
 
 type ColorSlot = u8;
 
-
 pub(crate) struct GlContainer {
     context: GlContext,
+
+    /// In order to set the current context we must have access to the instance
+    #[cfg(feature = "surfman")]
+    surfman_data: Option<(
+        Starc<RwLock<surfman::Device>>,
+        Starc<RwLock<surfman::Context>>,
+    )>,
```

Comment:
@kvark Is there any problem, do you think, with using `RwLock`'s to access the surfman `Device` and `Context` here?

The issue is that, in order to actually make the context current, you have to have mutable access to the surfman Device. Also, in the `Instance` you need mutable access to the Device to create contexts and surfaces. Locking was the only way I could think to allow both the `GLContainer` and the `Instance` have mutable access to the `Device` and `Context`.

Reply:
It should be fine, although we'll likely need `Arc` instead of `Starc` here, depending on the `Sync` guarantees of `surfman`. The most frequent operations in the API are issuing commands within render passes. Since GL doesn't have command buffers, we are recording the commands "in software", hence we don't need to access the GL context during this hot code. We'll need to access it only when doing device-level operations or submitting command buffers, which are not as frequent, and we can totally live with an `RwLock`.

Reply:
The `surfman::Context` and `Device` are `!Sync` and `!Send` so `Arc` didn't work. Will that be an issue.

Reply:
Hmm, it's certainly not as good as we want, but the main thing to address here is the divergence with the API. Having extra constraints is fine.